File: python/example/sort/sort_pure.py
import sys
from example.sort.service.pure.heapsort_service import HeapsortService
from example.sort.service.pure.quicksort_service import QuicksortService
from example.sort.tree.pure.node import Node

class Sort:

    @staticmethod
    def main(args):
        if len(args) < 1:
            print("Usage: <quick|heap> ...values")
            return 
        sort_algorithm = args[0]
        is_quick = sort_algorithm == "quick"
        is_heap = sort_algorithm == "heap"
        if not (is_quick or is_heap):
            print("Usage: <quick|heap> ...values")
            return
        args = args[1:]
        sort_service = QuicksortService()
        if is_heap:
            sort_service = HeapsortService()
        args = sort_service.sort(args)
        for arg in args:
            print(arg)

        a = Node(1, [])
        b = Node(2, [])
        c = Node(3, [a,b])
        d = Node(4, [])
        e = Node(5, [])
        f = Node(6, [d,e])
        g = Node(7, [c,f])

        list = [g]
        while len(list) > 0:
            new_list = []
            rest = False
            for el in list:
                if rest:
                    print(' ', end='')
                print(f'value: {el.value()}', end='')
                new_list.extend(el.children())
                rest = True
            print()
            list = new_list

if __name__ == "__main__":
    Sort.main(sys.argv[1:])

File: ts/example/sort/SortPure.ts
import { SortService } from './service/pure/SortService';
import { HeapsortService } from './service/pure/HeapsortService';
import { QuicksortService } from './service/pure/QuicksortService';
import { Node } from './tree/pure/Node';

function main() {
    let args = process.argv.slice(2);
    if (args.length < 1) {
        console.log("Usage: <quick|heap> ...values");
        return;
    }
    const sortAlgorithm = args[0];
    const isQuick = sortAlgorithm === "quick";
    const isHeap = sortAlgorithm === "heap";
    if (!(isQuick || isHeap)) {
        console.log("Usage: <quick|heap> ...values");
        return;
    }
    args = args.slice(1);
    let sortService : SortService = new QuicksortService();
    if (isHeap)
        sortService = new HeapsortService();
    args = sortService.sort(args);
    for (let arg of args) {
        console.log(arg);
    }

    const a = new Node(1, []);
    const b = new Node(2, []);
    const c = new Node(3, [a,b]);
    const d = new Node(4, []);
    const e = new Node(5, []);
    const f = new Node(6, [d,e]);
    const g = new Node(7, [c,f]);

    let list = [g]
    while (list.length > 0) {
        const newList = [];
        let rest = false;
        for (let el of list) {
            if (rest)
                process.stdout.write(' ')
            process.stdout.write(`value: ${el.value()}`);
            for (let child of el.children()) {
                newList.push(child);
            }
            rest = true;
        }
        process.stdout.write('\n');
        list = newList;
    }
}

if (require.main === module) {
    main();
}
